server/src/main/java/de/uftos/repositories/solver/timefold/constraints/ConstraintDefinitionFactory.java

```diff
@@ -0,0 +1,1073 @@
+    };
+  }
+
+  private static Function<List<ResourceTimefoldInstance>, Set<ResourceTimefoldInstance>>
+      convertResourceSet(AbstractSyntaxTreeDto ast, LinkedHashMap<String, ResourceType> params) {
+    if (ast.getToken() != UcdlToken.RESOURCE_SET) {
+      throw new IllegalStateException();
+    }
+    SetDto set = (SetDto) ast;
+
+    Function<List<ResourceTimefoldInstance>, ResourceTimefoldInstance> name;
+
+    if (set.setName().getToken() == UcdlToken.VALUE_REFERENCE) {
+      name = convertValueReference(set.setName(), params);
+    } else {
+      name = convertElement(set.setName(), params);
+    }
+
+    List<Function<
+        List<ResourceTimefoldInstance>,
+        Function<Set<ResourceTimefoldInstance>, Set<ResourceTimefoldInstance>>
+        >> modifiers = new ArrayList<>();
+
+    for (AbstractSyntaxTreeDto modifier : set.modifiers()) {
+      if (modifier.getToken() == UcdlToken.ATTRIBUTE) {
+        modifiers.add(convertAttribute(modifier, params));
+      } else {
+        modifiers.add(convertFilter(modifier, params));
+      }
+    }
+
+    return (parameters) -> {
+      Set<ResourceTimefoldInstance> setReference = new HashSet<>();
+      setReference.add(name.apply(parameters));
+
+      for (Function<
+            List<ResourceTimefoldInstance>,
+            Function<Set<ResourceTimefoldInstance>, Set<ResourceTimefoldInstance>>
+            > function : modifiers) {
+        setReference = function.apply(parameters).apply(setReference);
+      }
+      return setReference;
+    };
+  }
+
+  private static Function<List<ResourceTimefoldInstance>, Set<ResourceTimefoldInstance>>
+      convertNumberSet(AbstractSyntaxTreeDto ast, LinkedHashMap<String, ResourceType> params) {
+    if (ast.getToken() != UcdlToken.NUMBER_SET) {
+      throw new IllegalStateException();
+    }
+    ValueDto<Integer[]> set = (ValueDto<Integer[]>) ast;
+
+    Set<ResourceTimefoldInstance> resources = new HashSet<>();
+
+    for (Integer i : set.value()) {
+      resources.add(new Number(i));
+    }
+
+    return (parameters) -> resources;
+  }
+
+  private static Function<
+      List<ResourceTimefoldInstance>,
+      Function<Set<ResourceTimefoldInstance>, Set<ResourceTimefoldInstance>>
+      > convertAttribute(
+      AbstractSyntaxTreeDto ast, LinkedHashMap<String, ResourceType> params) {
+    if (ast.getToken() != UcdlToken.ATTRIBUTE) {
+      throw new IllegalStateException();
+    }
+
+    String attribute = ((ValueDto<String>) ast).value();
+
+    return switch (attribute) {
+      case "index" -> (parameters) -> (set) -> {
+        Set<ResourceTimefoldInstance> resultSet = new HashSet<>();
+        for (ResourceTimefoldInstance resource : set) {
+          if (resource.getResourceType() == ResourceType.LESSON) {
+            resultSet.add(new Number(((LessonTimefoldInstance) resource).getIndex()));
+          } else {
+            throw new IllegalStateException();
+          }
+        }
+        return resultSet;
```

Comment:
Might be worth moving all of those into their own functions as well

Reply:
I'll think about it.